sceptre/plan/actions.py

```diff
@@ -95,6 +96,7 @@ def create(self):
                 "%s - Create stack response: %s", self.stack.name, response
             )
 
+            self._note_response_time(response)
```

Comment:
Rather than this being on this class... I'm thinking it would make sense for this to be on the ConnectionManager itself. Basically, it records the timestamp of each response and makes it available as a property on it.

Reply:
This, actually, was not my first choice. The class, as written, stores this value as an instance attribute:

https://github.com/Sceptre/sceptre/blob/b73af2f341ff3b7b9ca4192858b69732fb26f195/sceptre/plan/actions.py#L787-L789

This value is "passed" to `_log_new_events` via this instance property. My first approach was to update the signature of `_wait_for_completion`:

```python
    def _wait_for_completion(self, timeout=0, start_time=None):
        # ...
        self.most_recent_event_datetime = start_time or datetime.now(tzutc()) - timedelta(seconds=3)
 ```

This broke several unit tests whose mocks weren't expecting the added parameter in the method calls. What I'm doing here resolves those but I'm not super happy with it either. If we're free to re-work things a bit, I'd argue that rather than an instance attribute, `most_recent_event_datetime` should be a stack variable in `_wait_for_completion` which is passed to `_log_new_events`.

`_log_new_events` would then return either:

- The datetime of the last event; or
- The datetime it was passed if there were no events.

This way we can eliminate some instance state which might cause issues down the line. We'll still need to update the signature for `_wait_for_completion`. Given this is an internal method and always invoked after a boto call, how would y'all feel about just passing the boto response?

```python
    def _wait_for_completion(self, timeout=0, boto_response=None):
```

Reply:
I think refactoring these private methods to be easier to test/more maintainable and overall reducing instance state sounds great to me and I'd definitely support that. As far as a stack variable... I don't think that's the appropriate place for them, though. If it's not held via the connection manager, passing parameters makes sense. Alternatively, using generators to maintain state in a really self-contained way can be a handy trick. Not sure if it'd apply here.

As far as the tests, it's not a secret that our tests need overhaul. We have several open Issues just for test improvement. As far as I'm concerned here, follow the "boyscout principle": Leave the code better than you found it. In this case, if the tests (which I recognize are overly rigid and poorly made) are fighting you, let's refactor/fix those tests to improve the quality of the tests and improve code coverage. You'll get 1000 Jon-points if you accomplish that without using `patch`. Don't go overboard and refactor things that are unrelated here. But for the things that touch the code you're introducing, by all means, be a boyscout.

Reply:
Sorry, what I meant by "stack variable" was "variable on the stack as opposed to the heap." Possibly not the best term in Python. I just mean, "plain old variable scoped within `_wait_for_completion` and passed as an argument to `_log_new_events`.

My approach in this commit was definitely less "boyscout" and more "touch as little as possible." Let me do some actual refactoring and see if I can earn those Jon-points.